### PR TITLE
[IMP] account_edi_proxy_client,l10n_it_edi: remove edi format dependency

### PR DESCRIPTION
**Pull request by aliyatastemirova**
Edi format is going to be removed eventually - `account_edi_proxy_client` should not depend on it and be more flexible in general.

This commit:
- Removes edi_format field from `account_edi_proxy_client.user` model. We still have to add them in `l10n_it_edi` as it still heavily relies on edi format for now.
- Adds a way to have a unique edi user per company, proxy user, edi mode combination.

This is a necessary step before adding peppol, which will also depend on the account_edi_proxy_client

also see: https://github.com/odoo/iap-apps/pull/580

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account_edi_proxy_client/models/account_edi_proxy_user.py [hunk 1]
```diff
@@ -45,28 +42,73 @@ class AccountEdiProxyClientUser(models.Model):
...
+            "l10n_it_edi": {
+                "prod": "https://l10n-it-edi.api.odoo.com",
+                "test": "https://iap-services-test.odoo.com",
+            },
+        }
+        return proxy_urls
+
+    def _get_edi_mode(self):
+        edi_mode = self.env['ir.config_parameter'].sudo().get_param('account_edi_proxy_client.demo', False)
+        return 'prod' if edi_mode in ['prod', False] else 'test' if edi_mode == 'test' else 'demo'
```
Comment (smetl) [inline addons/account_edi_proxy_client/models/account_edi_proxy_user.py:90]:
```suggestion
        return 'prod' if edi_mode in ('prod', False) else 'test' if edi_mode == 'test' else 'demo'
```

File: addons/account_edi_proxy_client/models/account_edi_proxy_user.py [hunk 2]
```diff
@@ -45,28 +42,73 @@ class AccountEdiProxyClientUser(models.Model):
...
+            user.edi_mode = user._get_edi_mode()
+
+    def _get_proxy_urls(self):
+        proxy_urls = {
+            "l10n_it_edi": {
+                "prod": "https://l10n-it-edi.api.odoo.com",
+                "test": "https://iap-services-test.odoo.com",
+            },
+        }
+        return proxy_urls
```
Comment (smetl) [inline addons/account_edi_proxy_client/models/account_edi_proxy_user.py:86]:
```suggestion
        return {
            "l10n_it_edi": {
                "prod": "https://l10n-it-edi.api.odoo.com",
                "test": "https://iap-services-test.odoo.com",
            },
        }
```

File: addons/account_edi_proxy_client/models/account_edi_proxy_user.py [hunk 3]
```diff
@@ -45,28 +42,73 @@ class AccountEdiProxyClientUser(models.Model):
...
-    def _get_server_url(self):
-        return DEFAULT_TEST_SERVER_URL if self._get_demo_state() == 'test' else self.env['ir.config_parameter'].sudo().get_param('account_edi_proxy_client.edi_server_url', DEFAULT_SERVER_URL)
+    def _get_proxy_user(self, company, proxy_type, edi_mode):
+        '''Returns the proxy_user associated with this edi format.
+        '''
+        self.ensure_one()
+        return company.account_edi_proxy_client_ids.filtered(lambda u: u.proxy_type == proxy_type and u.edi_mode == edi_mode)
+
+    def _get_proxy_identification(self, company):
+        '''Returns the key that will identify company uniquely for this edi format (for example, the vat)
```
Comment (smetl) [inline addons/account_edi_proxy_client/models/account_edi_proxy_user.py:107]: change the comment about edi format

File: addons/l10n_it_edi/models/account_edi_proxy_user.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    edi_format_id = fields.Many2one('account.edi.format', required=True)
+    edi_format_code = fields.Char(related='edi_format_id.code', readonly=True)
+
+    _sql_constraints = [
+        ('unique_edi_identification_per_for', 'unique(edi_identification, edi_format_id)', 'This edi identification is already assigned to a user'),
+    ]
+
+    def _prepare_vals_edi_proxy_client(self, response, company, proxy_type, edi_identification, private_pem):
+        vals = super()._prepare_vals_edi_proxy_client(response, company, proxy_type, edi_identification, private_pem)
+        vals['edi_format_id'] = self.env.ref('l10n_it_edi.edi_fatturaPA').id
```
Comment (smetl) [inline addons/l10n_it_edi/models/account_edi_proxy_user.py:18]: I don't understand why you still need that

Comment (aliyatastemirova) [reply addons/l10n_it_edi/models/account_edi_proxy_user.py:18]: I just don't want to change anything about the l10n_it_edi flow at the moment and I think they still need edi_format_id

File: addons/account_edi_proxy_client/views/account_edi_proxy_user_views.xml [hunk 5]
```diff
@@ -10,7 +10,6 @@
                         <group>
                             <group>
                                 <field name="company_id" groups="base.group_multi_company" invisible="1"/>
-                                <field name="edi_format_id" readonly="1" options="{'no_open': True}"/>
```
Comment (lordkrandel) [inline addons/account_edi_proxy_client/views/account_edi_proxy_user_views.xml:13]: Can we add the `proxy_type` ? If we add PEPPOL we'll have to tell one from the other

File: addons/account_edi_proxy_client/models/account_edi_proxy_user.py [hunk 6]
```diff
@@ -45,28 +42,63 @@ class AccountEdiProxyClientUser(models.Model):
...
+        edi_mode = self.env['ir.config_parameter'].sudo().get_param('account_edi_proxy_client.demo', False)
+        return 'prod' if edi_mode in ('prod', False) else 'test' if edi_mode == 'test' else 'demo'
 
-    def _get_server_url(self):
-        return DEFAULT_TEST_SERVER_URL if self._get_demo_state() == 'test' else self.env['ir.config_parameter'].sudo().get_param('account_edi_proxy_client.edi_server_url', DEFAULT_SERVER_URL)
+    def _get_server_url(self, proxy_type=None):
+        proxy_type = proxy_type or self.proxy_type
+        edi_mode = self._get_edi_mode()
+        proxy_urls = self._get_proxy_urls()
+        if not (proxy_urls.get(proxy_type) and proxy_urls[proxy_type].get(edi_mode)):
```
Comment (lordkrandel) [inline addons/account_edi_proxy_client/models/account_edi_proxy_user.py:85]: Think we can let this traceback ... it's not a configuration, it's based on code. 
I'd rather have a test that ensures the configuration is correct, then a UserError at runtime.
But I'd like to hear your opinion on this :)

File: addons/account_edi_proxy_client/models/account_edi_proxy_user.py [hunk 7]
```diff
@@ -45,28 +42,62 @@ class AccountEdiProxyClientUser(models.Model):
...
+            ('demo', 'Demo mode'),
+        ],
+        string='EDI operating mode',
+        compute='_compute_edi_mode', readonly=False, store=True,
+    )
 
     _sql_constraints = [
         ('unique_id_client', 'unique(id_client)', 'This id_client is already used on another user.'),
-        ('unique_edi_identification_per_for', 'unique(edi_identification, edi_format_id)', 'This edi identification is already assigned to a user'),
+        ('unique_edi_identification', 'unique(edi_identification, proxy_type, edi_mode)', 'This edi identification is already assigned to a user'),
```
Comment (smetl) [inline addons/account_edi_proxy_client/models/account_edi_proxy_user.py:62]: Idk if you need something in the upgrade script to add the new constraint

File: addons/l10n_it_edi/models/res_config_settings.py [hunk 8]
```diff
@@ -62,7 +62,7 @@ def button_create_proxy_user(self):
         if not edi_identification:
             return
 
-        self.env['account_edi_proxy_client.user']._register_proxy_user(self.company_id, fattura_pa, edi_identification)
+        self.env['account_edi_proxy_client.user']._register_proxy_user(self.company_id, 'l10n_it_edi', edi_identification)
```
Comment (lordkrandel) [inline addons/l10n_it_edi/models/res_config_settings.py:65]: We cannot change this unless we change this 😅  
image

Comment (lordkrandel) [reply addons/l10n_it_edi/models/res_config_settings.py:65]: Or we can put `('edi_format_code', 'in', ('fattura_pa', 'l10n_it_edi'))` and have a transition period

File: addons/l10n_it_edi/models/account_edi_format.py [hunk 9]
```diff
@@ -890,13 +890,13 @@ def _l10n_it_post_invoices_step_2(self, invoices):
         to_return = {}
         company = invoices.company_id
 
-        proxy_user = self._get_proxy_user(company)
+        proxy_user = self.env['account_edi_proxy_client.user']._get_proxy_users(company, 'l10n_it_edi')
```
Comment (aliyatastemirova) [inline addons/l10n_it_edi/models/account_edi_format.py:893]: at the moment we don't allow creating a separate user per edi mode, so this one can only return one edi user.

File: addons/l10n_it_edi/models/account_edi_proxy_user.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+from odoo import fields, models
+
+
+class AccountEdiProxyClientUser(models.Model):
+    _inherit = 'account_edi_proxy_client.user'
+
+    proxy_type = fields.Selection(
+        selection_add=[('l10n_it_edi', 'Italian EDI')],
+        compute='_compute_proxy_type', readonly=False, store=True,
```
Comment (smetl) [inline addons/l10n_it_edi/models/account_edi_proxy_user.py:12]: In `account_edi_proxy_client`:
```
proxy_type = fields.Selection(
    selection=[],
    string="Proxy type",
    compute='_compute_proxy_type',
    readonly=False,
    store=True,
)

@api.depends('company_id')
def _compute_proxy_type(self):
    for user in self:
        user.proxy_type = False
```
In `l10n_it_edi`:
```
def _compute_proxy_type(self):
    # Extends account_edi_proxy_client
    super()._compute_proxy_type()
    for user in self:
        if user.company_id_country_code == 'IT':
            user.proxy_type = 'l10n_it_edi'
```

**Comment by aliyatastemirova** [comment]
> Missing dependency of `l10n_it_edi` on `account_edi_proxy_client`.

what do you mean? `l10n_it_edi` has a dependency on `account_edi_proxy_client` already. And `account_edi_proxy_client` does not depend on `l10n_it_edi`.

**Review by smetl — APPROVED** [review]
*awaiting l10n build*

**Comment by smetl** [comment]
@robodoo r+

**Comment by lordkrandel** [comment]
As the setting will be company-specific, please add the related icon 😉 
You can find it at this line in my old PR to copy paste 😄 

https://github.com/odoo/odoo/pull/110249/files#diff-97a0f7d9e3336dfe2fc22abaa394ac4bc6bfddd9a0417a55401e12092c0b9976R16

**Comment by smetl** [comment]
@robodoo r+

**Pull request closed without merging**